legacy/src/dnn_backend/dnn_backend_miopen.cpp.inc

```diff
@@ -0,0 +1,1236 @@
+                                          size_t ndims,
+                                          int* dims,
+                                          int* strides)
+{
+    DISTCONV_CHECK_MIOPEN(
+        miopenSetTensorDescriptor(desc, dt, ndims, dims, strides));
+}
+
+void GPUDNNBackend::set_tensor_descriptor(TensorDescriptor_t const& desc,
+                                          DataType_t const& dt,
+                                          std::vector<int> const& dims,
+                                          std::vector<int> const& strides)
+{
+    assert_eq(dims.size(), strides.size());
+    set_tensor_descriptor(desc,
+                          dt,
+                          dims.size(),
+                          const_cast<int*>(dims.data()),
+                          const_cast<int*>(strides.data()));
+}
+
+auto GPUDNNBackend::make_filter_descriptor() -> TensorDescriptor_t
+{
+    return make_tensor_descriptor();
+}
+
+void GPUDNNBackend::destroy_filter_descriptor(TensorDescriptor_t const& desc)
+{
+    destroy_tensor_descriptor(desc);
+}
+
+void GPUDNNBackend::set_filter_descriptor(FilterDescriptor_t const& desc,
+                                          DataType_t const& dt,
+                                          size_t ndims,
+                                          int const* dims)
+{
+    set_filter_descriptor(desc, dt, std::vector<int>{dims, dims + ndims});
+}
+
+void GPUDNNBackend::set_filter_descriptor(FilterDescriptor_t const& desc,
+                                          DataType_t const& dt,
+                                          std::vector<int> const& dims)
+{
+    auto const strides = get_fully_packed_strides(dims);
+    set_tensor_descriptor(desc, dt, dims, strides);
+}
+
+int GPUDNNBackend::get_tensor_rank(TensorDescriptor_t const& desc)
+{
+    int num_dims = -1;
+    DISTCONV_CHECK_MIOPEN(miopenGetTensorDescriptorSize(desc, &num_dims));
+    return num_dims;
+}
+
+int GPUDNNBackend::get_tensor_dimension(TensorDescriptor_t const& desc, int d)
+{
+    int const num_dims = get_tensor_rank(desc);
+    d = d < 0 ? num_dims + d : d;
+    assert_always(d < num_dims);
+
+    miopenDataType_t dt;
+    std::vector<int> dims, strides;
+    dims.reserve(num_dims);
+    strides.reserve(num_dims);
+    DISTCONV_CHECK_MIOPEN(
+        miopenGetTensorDescriptor(desc, &dt, dims.data(), strides.data()));
+    return dims[num_dims - d - 1];
+}
+
+void GPUDNNBackend::set_tensor_dimension(TensorDescriptor_t& desc, int d, int n)
+{
+    int const num_dims = get_tensor_rank(desc);
+    d = d < 0 ? num_dims + d : d;
+    assert_always(d < num_dims);
+
+    miopenDataType_t dt;
+    std::vector<int> dims, strides;
+    dims.reserve(num_dims);
+    strides.reserve(num_dims);
+
+    DISTCONV_CHECK_MIOPEN(
+        miopenGetTensorDescriptor(desc, &dt, dims.data(), strides.data()));
+    dims[num_dims - d - 1] = n;
+    // FIXME (TRB): Need to recompute strides??
```

Comment:
likely yes. I think we must or this will fail in unexpected places

Reply:
This is a copy-paste `FIXME`, but I concluded that there's not enough information to be able to compute the strides correctly. This could be a resize within the "bounding extents", for example.

There are also other issues with this API that make it rather confusing (e.g., `d` is relative to WHDCN ordering of the tensor indices that DistConv tensors use (hence the `dims[num_dims - d - 1]` computation)). It will be strongly reconsidered in future refactoring efforts.

Reply:
Fair enough